### PR TITLE
Implement visibility notifier and grid unloading

### PR DESCRIPTION
**Pull request by AGandrup**
Alright, here we go again. This is my new (and most likely last) attempt at porting TC's visibility and grid unloading functionality to AC. First attempt didn't go quite as planned, mainly because I went way out of scope for that PR (at the time I didn't realize that), which only resulted in shitload of bugs and crashed. This time, I have went over the entire thing once more and tried to be more thorough with testing and analyzing each change. I have also been extra careful not to go out of scope and only made change which I deemed absolutely necessary - though, with that said, I've still kept some minor refactoring here and there but I do consider reverting those to make the PR less overwhelming and easier to review.
 
## Changes Proposed:
The motivation behind this PR is:
- Better performance
- Implement grid unloading
- Implement visibility notifier
- See creatures and gameobjects from a larger distance
- More alignment with TC.

## Issues Addressed:

A lot of known (reported) issues as well as other that I've found, which were introduced in the first attempt, should all have been resolved now. This includes but not limited to:
- https://github.com/azerothcore/azerothcore-wotlk/issues/17580
- https://github.com/azerothcore/azerothcore-wotlk/issues/17595
- https://github.com/azerothcore/azerothcore-wotlk/issues/17575
- https://github.com/chromiecraft/chromiecraft/issues/6299

**The following issue I haven't been able to reproduce and therefore haven't been able to confirm yet:**
- https://github.com/azerothcore/azerothcore-wotlk/issues/17605
- https://github.com/azerothcore/azerothcore-wotlk/issues/17439
- https://github.com/azerothcore/azerothcore-wotlk/issues/17434

I believe (hope) that these issues has been resolved as well. As for the high diff/high cpu issue I need someone with a high populated server to test it for me (Looking at you @mpfans 👀). I welcome anyone to test it 🙏 

## Tests Performed:
- Builds
- Tested in game
- Analyzed via Grafana and Windows Performance Monitor

## How to Test the Changes:

In terms of testing performance I recommend using Grafana, any CPU/Memory monitor and/or perhaps [VS performance utility](https://learn.microsoft.com/da-dk/previous-versions/visualstudio/visual-studio-2015/profiling/beginners-guide-to-performance-profiling?view=vs-2015&redirectedfrom=MSDN).

**Keep in mind that grid unloading was never a thing on AC and must now manually be disabled in worldserver.conf. It is enabled by default.**

## Known Issues and TODO List:

- None.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by AnchyDev** [comment]
i believe in u

**Comment by AGandrup** [comment]
Here is a few metrics and comparisons:

1) Master branch with preloading grids disabled. (Metrics disabled). Avg. CPU usage: ~37%
image
Metrics enabled:
image
image

2) Master branch with preloading grids enabled. (Metrics disabled). Avg. CPU usage: ~90%
image
Metrics enabled:
image
image

3) PR with preloading grids disabled and unloading grids enabled. (Metrics disabled). Avg. CPU usage: ~1%
image
Metrics enabled:
image
image

4) PR with preloading grids disabled and unloading grids disabled. This is equivalent to how AC works today, out of the box. (Metrics disabled). Avg. CPU usage: ~5%
image
Metrics enabled:
image
image

5) PR with preloading grids enabled. (Metrics disabled) Avg. CPU usage: ~63%
image
Metrics enabled:
image
image

Each measurement is captured over a rougly 30 min timeframe with only 1 player online so just keep that in mind before jumping to any conclusion. Throughout the measurement I was fast traveling through continents to trigger grid loading, active objects and what not - took the same route every time. But with that said, the server was basically idle, so again, I need someone with a high pop server to test.

Still interesting to see that roughly 30% CPU usage was cut off on "idle" server. Hope it stays that way with high pop server 😛

**Comment by AGandrup** [comment]
Good question lol. It is ready for review.

**Comment by Kitzunu** [comment]
Need to resolve conflict

**Comment by mpfans** [comment]
‘‘‘Automatic merge failed; fix conflicts and then commit the result.’’’

**Comment by AGandrup** [comment]
> `Automatic merge failed; fix conflicts and then commit the result.`

Fixed now, sorry for the wait.

**Comment by AGandrup** [comment]
To testers and just a general headsup: I forgot that I had applied the following two database fixes, which helps with performance as well:
https://github.com/azerothcore/azerothcore-wotlk/pull/17886
https://github.com/azerothcore/azerothcore-wotlk/pull/17888

Especially the first one.

**Comment by Kitzunu** [comment]
@mpfans @pklloveyou @azerothcore/testers Could anyone test this in a production environment please

**Comment by Kitzunu** [comment]
This PR should potentially fix https://github.com/azerothcore/azerothcore-wotlk/issues/3269 right?

**Comment by mpfans** [comment]
```Update time diff spike: 105. Update time diff average: 46. Players online: 510.```

**Comment by AGandrup** [comment]
Btw spikes seems to be caused by game events starting and stopping. It is an issue on master as well.

**Comment by mpfans** [comment]
I need to wait for more players to see the difference

**Comment by AGandrup** [comment]
Ye understandable, I'll keep my fingers crossed 😛 🤞

**Comment by AGandrup** [comment]
> This PR should potentially fix #3269 right?

Not entirely. Only when the instance is unloaded (default after 5 min) npcs are moved to their correct positions.
Im gonna take a guess and say this issue was most likely fixed in the big overhaul on creature and gob spawn system.

**Comment by mpfans** [comment]
```Update time diff spike: 185. Update time diff average: 110. Players online: 987.```

**Comment by AGandrup** [comment]
@mpfans do you have Grafana enabled? And can you share your configs?

**Comment by mpfans** [comment]
> @mpfans do you have Grafana enabled? And can you share your configs?

Not enabled, please receive configuration files in Discord

**Comment by AGandrup** [comment]
> > @mpfans do you have Grafana enabled? And can you share your configs?
> 
> Not enabled, please receive configuration files in Discord

Thanks! It might be worth enabling Grafana to troubleshoot easier. However I can see that you have Grid Unload enabled, so to being with, if you continue to have high update time diff you should consider disabling it. Grid Unload is enabled by default but once you get more players then, depending on hardware, it is often recommended to disable it.

Try that first and if you continue to have high update time diff, try enable preloading of base maps.

**Comment by mpfans** [comment]
> > > @mpfans do you have Grafana enabled? And can you share your configs?
> > 
> > 
> > Not enabled, please receive configuration files in Discord
> 
> Thanks! It might be worth enabling Grafana to troubleshoot easier. However I can see that you have Grid Unload enabled, so to being with, if you continue to have high update time diff you should consider disabling it. Grid Unload is enabled by default but once you get more players then, depending on hardware, it is often recommended to disable it.
> 
> Try that first and if you continue to have high update time diff, try enable preloading of base maps.

```Update time diff spike: 581. Update time diff average: 663. Players online: 1905.```
I am about to make the following settings and restart the server
```Grid Unload =0 
preloading of base map = 1```

**Comment by mpfans** [comment]
After restarting the server
```Update time diff spike: 214. Update time diff average: 183. Players online: 1417.```

**Comment by AGandrup** [comment]
I'll have to look through your config when I get home. But I think it could help a lot if you installed Grafana - that way it might be easier to tell whats happening.

**Comment by AGandrup** [comment]
> > I'll have to look through your config when I get home. But I think it could help a lot if you installed Grafana - that way it might be easier to tell whats happening.
> 
> `Update time diff spike: 633. Update time diff average: 479. Players online: 2408.`
> 
> I have used Grafana before, but only the first 5 minutes of running have data, and then there is no more data, which may be due to my large amount of data

That should be fixed now.

**Comment by mpfans** [comment]
And I found that this PR(https://github.com/azerothcore/azerothcore-wotlk/pull/17102) also has a significant impact on Update time diff, so when I upgrade to the latest core, I will locally reverse this PR(https://github.com/azerothcore/azerothcore-wotlk/pull/17102)

File: src/server/scripts/Northrend/IcecrownCitadel/boss_sindragosa.cpp [hunk 1]
```diff
@@ -1549,6 +1554,8 @@ class npc_rimefang : public CreatureScript
             if (point == POINT_FROSTWYRM_LAND)
             {
                 me->setActive(false);
+                me->SetFarVisible(false);
+                me->SetFarVisible(false);
```
Comment (IntelligentQuantum) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_sindragosa.cpp:1558]: Do we need two SetFar ?

**Comment by moncat2005** [comment]
> > I am using the latest core and not testing this PR 
> `Update time diff: 74. Players online: 1545.`
> 
> `Update time diff: 90. Players online: 1845.`
> 
> `Update time diff: 132. Players online: 2121.`

@mpfans
Is your current WorldServer service online volume and UTD score based on the latest code? 
What issuse numbers have been reversed in total?

**Comment by AGandrup** [comment]
Quick update: I have not abandoned this, there's just some other things taking my attention.
I have done some debugging and I see where all the resources go and I also see why it isn't a problem on master. To make a long story short, on master we have this dynamic visibility class which throttles object relocation thus object updates which is what eats a lot of resources because of all the heavy calculations done there. These calculations seems to be more optimized on TC so it isn't as big of a problem (or a problem at all) there, thus no need to throttle.
So I've been debating with myself whether I should keep the throttling or try fix the root of the problem. The latter would be preferred I guess but I am not sure how much value is gained from it and how long it would take. Throttling is quick but also a dirty hack(?) I guess, but AC has been running for years with it soo, idk.
It also got me thinking, if I should keep it then I might as well just improve on what we got instead of trying to rework the entire thing. I haven't really decided yet.

Any thoughts? @Kitzunu perhaps?

**Comment by Kitzunu** [comment]
I would say instead of doing everything at once. do it bit by bit

**Comment by Nyeriah** [comment]
Marking as abandoned since this PR hasn't received any updates in a long while. If there's interest in continuing this PR please leave a message and we will reopen it

**Pull request closed without merging**